crates/orderbook/src/order_quoting.rs

```diff
@@ -332,19 +368,202 @@ pub trait OrderQuoteCalculating: Send + Sync {
+    async fn save(&self, data: QuoteData) -> Result<QuoteId>;
+
+    /// Retrieves an existing quote by ID.
+    async fn get(&self, id: QuoteId, expiration: DateTime<Utc>) -> Result<Option<QuoteData>>;
+
+    /// Retrieves an existing quote by ID.
+    async fn find(
+        &self,
+        parameters: QuoteSearchParameters,
+        expiration: DateTime<Utc>,
+    ) -> Result<Option<(QuoteId, QuoteData)>>;
+}
+
+#[cfg_attr(test, mockall::automock)]
+pub trait Now: Send + Sync {
+    fn now(&self) -> DateTime<Utc>;
+}
+
+impl<F> Now for F
+where
+    F: Fn() -> DateTime<Utc> + Send + Sync,
+{
+    fn now(&self) -> DateTime<Utc> {
+        (self)()
+    }
+}
+
+/// How long a quote remains valid for.
+const QUOTE_VALIDITY_SECONDS: i64 = 60;
+
+/// An order quoter implementation that relies
+pub struct OrderQuoter2 {
+    price_estimator: Arc<dyn PriceEstimating>,
+    native_price_estimator: Arc<dyn NativePriceEstimating>,
+    gas_estimator: Arc<dyn GasPriceEstimating>,
+    storage: Arc<dyn QuoteStoring>,
+    now: Arc<dyn Now>,
+}
+
+impl OrderQuoter2 {
+    async fn compute_quote_data(
+        &self,
+        parameters: &QuoteParameters,
+    ) -> Result<QuoteData, CalculateQuoteError> {
+        let expiration = self.now.now() + chrono::Duration::seconds(QUOTE_VALIDITY_SECONDS);
+
+        let trade_query = parameters.to_price_query();
+        let (gas_estimate, trade_estimate, sell_token_price) = futures::try_join!(
+            self.gas_estimator
+                .estimate()
+                .map_err(PriceEstimationError::from),
+            single_estimate(self.price_estimator.as_ref(), &trade_query),
+            native_single_estimate(self.native_price_estimator.as_ref(), &parameters.sell_token),
+        )?;
+
+        let fee_parameters = FeeParameters {
+            gas_amount: trade_estimate.gas as _,
+            gas_price: gas_estimate.effective_gas_price(),
+            sell_token_price,
+        };
+        // TODO(nlordell): Apply subsidies!
+        let fee_amount = fee_parameters.amount_in_sell_token();
+
+        let (sell_amount, buy_amount) = match &parameters.side {
+            OrderQuoteSide::Sell {
+                sell_amount:
+                    SellAmount::BeforeFee {
+                        value: sell_amount_before_fee,
+                    },
+            } => {
+                let sell_amount = sell_amount_before_fee.saturating_sub(fee_amount);
+                if sell_amount == U256::zero() {
+                    // We want a sell_amount of at least 1!
+                    return Err(CalculateQuoteError::SellAmountDoesNotCoverFee(fee_amount));
+                }
+
+                let buy_amount = match trade_estimate.out_amount.checked_mul(sell_amount) {
+                    Some(product) => product / sell_amount_before_fee,
+                    // If we overflow when computing the product, use a different
+                    // less precise method that avoids overflows.
```

Comment:
Could also multiply in U512 but not sure if worth.

Reply:
Nice idea!

Reply:
Made the code nicer, so implemented this.